File: java_refactoring/src/main/java/org/example/solution/farmcreator/CropsFarmCreator.java
package org.example.solution.farmcreator;

import org.example.solution.FarmType;

public class CropsFarmCreator extends FarmCreator {


    public CropsFarmCreator(String farmName) {
        super(farmName, FarmType.fromString("crops"));
    }

    @Override
    public int getValueToAdd() {
        return 1000;
    }

    //farmName should be passed in a constructor
    @Override
    public void createLandType() {
        System.out.println("Creating land for " + this.farmName);
        for (int count = 0; count < FarmCreator.ADD_LAND_COUNT; count++) {
            this.farm.append( "crops");
            System.out.println("adding some crops " + this.farmName); // adding farmName to the log would make the log more informative
            System.out.println("water crops " + this.farmName); // adding farmName to the log would make the log more informative
            System.out.println("collect seeds " + this.farmName ); // adding farmName to the log would make the log more informative
            this.numCreated += 2;
        }
    }

    @Override
    public void addPlants() {
        System.out.println("Planting clover " + this.farmName);
    }



}

File: typescript_refactoring/example/solution/CropsFarmCreator.ts
import { FarmCreator } from "./FarmCreator";
import { FarmType } from "./FarmType";

export class CropsFarmCreator extends FarmCreator {
    public constructor(farmName: string) {
        super(farmName, FarmType.Crops);
    }

    public addPlants(): void {
        console.log(`Planting clover for farm ${this.farmName}`);
    }

    protected createLandType(): void {
        this.addFarmPart("crops crops");
        console.log(`Adding some crops for farm ${this.farmName}`);
        console.log(`Water crops for farm ${this.farmName}`);
        console.log(`Collect seeds for farm ${this.farmName}`);
        this.numCreated += 2;
    }

    protected getValueToAdd(): number {
        return 1000;
    }
}
